app/bundles/CampaignBundle/Entity/LeadEventLogRepository.php

```diff
@@ -169,8 +170,16 @@ public function getUpcomingEvents(array $options = null): array
         }
 
         if (isset($options['type'])) {
-            $query->andwhere('e.type = :type')
-                  ->setParameter('type', $options['type']);
+            $eventIdQuery = $this->_em->getConnection()->createQueryBuilder();
+            $eventIdQuery->select('e.id')
+                ->from(MAUTIC_TABLE_PREFIX.'campaign_events', 'e')
+                ->where($eventIdQuery->expr()->eq('e.type', ':type'))
+                ->setParameter('type', $options['type']);
+
+            $eventIds = $eventIdQuery->execute()->fetchFirstColumn();
+            if (!empty($eventIds)) {
+                $query->andWhere($query->expr()->in('ll.event_id', $eventIds));
+            }
```

Comment:
I think that this is not necessary, if you have like 8k send email events in your database it will create a large query that contains all the IDs. I've got similar time results when testing without this change.